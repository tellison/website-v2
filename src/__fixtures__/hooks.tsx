import {
  Binary,
  ContributorApiResponse,
  Contributor,
  EventAPI,
  LatestTemurin,
  MarketplaceRelease,
  MockTemurinFeatureReleaseAPI,
  MockTemurinReleaseAPI,
  News,
  NewsResponse,
  ReleaseAsset,
  TemurinReleases,
} from '../hooks';

export const createRandomTemurinRelease = (installer): ReleaseAsset => ({
  os: 'os_mock',
  architecture: 'arch_mock',
  release_link: new URL('https://release_link_mock'),
  platform_name: 'platform_name_mock',
  release_name: 'release_name_mock',
  release_date: new Date(Date.UTC(2020, 0, 1)),
  binaries: [
    {
      type: 'type_mock',
      link: new URL('https://link_mock'),
      checksum: 'checksum_mock',
      size: 0,
      extension: 'extension_mock',
      installer_link: installer ? new URL('https://installer_link_mock') : undefined,
      installer_checksum: installer ? 'installer_checksum_mock' : undefined,
      installer_size: installer ? 0 : undefined,
      installer_extension: installer ? 'installer_extension_mock' : undefined,
    },
  ],
})

export const createRandomMarketplaceRelease = (installer, id): MarketplaceRelease => ({
  release_name: 'release_name_mock',
  vendor: 'vendor_mock',
  binary: {
    os: 'os_mock',
    architecture: 'arch_mock',
    distribution: 'distribution_mock',
    image_type: 'type_mock',
    jvm_impl: 'jvm_impl_mock',
    timestamp: new Date(Date.UTC(2020, 0, 1)),
    scm_ref: 'scm_ref_mock',
    openjdk_scm_ref: 'openjdk_scm_ref_mock',
    aqavit_results_link: new URL('https://aqavit_results_link_mock'),
    tck_affidavit_link: new URL('https://tck_affidavit_link_mock'),
    package: {
      name: 'name_mock.tar.gz',
      link: new URL('https://link_mock'),
      sha265sum: `sha265sum_mock${id}`,
      sha256sum_link: new URL('https://sha256sum_link_mock'),
      signature_link: new URL('https://signature_link_mock'),
    },
    installer: installer
      ? [
          {
            name: 'installer_name_mock.msi',
            link: new URL('https://installer_link_mock'),
            sha265sum: 'installer_sha265sum_mock',
            sha256sum_link: new URL('https://installer_sha256sum_link_mock'),
            signature_link: new URL('https://installer_signature_link_mock'),
          },
        ]
      : undefined,
  },
  version: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  }
})

export const mockLatestTemurin = (installer): LatestTemurin => ({
  download_count: 0,
  id: 'id_mock',
  release_link: new URL('https://release_link_mock'),
  release_name: 'release_name_mock',
  release_type: 'release_type_mock',
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: 'vendor_mock',
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  },
  binaries: [
    {
      os: 'os_mock',
      architecture: 'arch_mock',
      image_type: 'type_mock',
      heap_size: 'heap_size_mock',
      download_count: 0,
      jvm_impl: 'jvm_impl_mock',
      package: {
        name: 'name_mock.tar.gz',
        link: new URL('https://link_mock'),
        checksum: 'sha265sum_mock',
        checksum_link: new URL('https://sha256sum_link_mock'),
        metadata_link: new URL('https://metadata_link_mock'),
        signature_link: new URL('https://signature_link_mock'),
        size: 0,
        download_count: 0,
      },
      installer: installer
      ?
        {
          name: 'installer_name_mock.msi',
          link: new URL('https://installer_link_mock'),
          checksum: 'installer_sha265sum_mock',
          checksum_link: new URL('https://installer_sha256sum_link_mock'),
          metadata_link: new URL('https://installer_metadata_link_mock'),
          signature_link: new URL('https://installer_signature_link_mock'),
          size: 0,
          download_count: 0,
        }
      : undefined,
    }
  ]
})

export const mockNewsAPI = (): NewsResponse => ({
  news: [
    {
      title: 'news_title_mock',
      body: 'body_mock',
      id: 'id_mock',
      link: new URL('https://link_mock'),
      date: new Date(Date.UTC(2020, 0, 1)),
    }
  ],
  pagination: {
    page: 1,
    pagesize: 5,
    result_start: 1,
    result_end: 5,
    result_size: 5,
    total_result_size: 16
  }
});

export const mockEventsAPI = (): EventAPI => ({
  events: [
    {
      title: 'events_title_mock',
      id: 'id_mock',
      infoLink: new URL('https://link_mock'),
      date: new Date(Date.UTC(2020, 0, 1)),
    }
  ]
});

export const createMockTemurinReleaseAPI = (installer): MockTemurinReleaseAPI => ({
  release_link: new URL('https://release_link_mock'),
  release_name: 'release_name_mock',
  vendor: 'vendor_mock',
  version: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  },
  binary: {
    architecture: 'arch_mock',
    project: 'project_mock',
    scm_ref: 'scm_ref_mock',
    download_count: 0,
    updated_at: new Date(Date.UTC(2020, 0, 1)),
    os: 'os_mock',
    image_type: 'jdk',
    jvm_impl: 'jvm_impl_mock',
    heap_size: 'heap_size_mock',
    package: {
      name: 'name_mock.tar.gz',
      link: new URL('https://link_mock'),
      checksum: 'checksum_mock',
      checksum_link: new URL('https://checksum_link_mock'),
      metadata_link: new URL('https://metadata_link_mock'),
      signature_link: new URL('https://signature_link_mock'),
      size: 0,
      download_count: 0,
    },
    installer: installer
      ? {
          name: 'installer_name_mock.msi',
          link: new URL('https://installer_link_mock'),
          checksum: 'installer_checksum_mock',
          checksum_link: new URL('https://installer_checksum_link_mock'),
          metadata_link: new URL('https://installer_metadata_link_mock'),
          signature_link: new URL('https://installer_signature_link_mock'),
          size: 0,
          download_count: 0,
        }
      : undefined,
  }
});

export const createMockTemurinFeatureReleaseAPI = (installer): MockTemurinFeatureReleaseAPI => ({
  id: 'id_mock',
  download_count: 0,
  release_name: 'release_name_mock',
  release_link: new URL('https://release_link_mock'),
  release_type: 'release_type_mock',
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: 'vendor_mock',
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  },
  binaries: [
    {
      os: 'os_mock',
      architecture: 'arch_mock',
      image_type: 'jdk',
      package: {
        name: 'name_mock.tar.gz',
        checksum: 'checksum_mock',
        link: new URL('https://link_mock'),
        size: 0,
      },
      installer: installer
        ? {
          name: 'installer_name_mock.msi',
          checksum: 'installer_checksum_mock',
          link: new URL('https://installer_link_mock'),
          size: 0,
        }
      : undefined,
    }
  ]
});

export const createRandomTemurinReleases = (installer): TemurinReleases => ({
  release_name: 'release_name_mock',
  release_link: new URL('https://release_link_mock'),
  source_url: new URL('https://source_url_mock'),
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  platforms: {
    'platform_mock': {
      assets: [
        {
          os: 'os_mock',
          architecture: 'architecture_mock',
          type: 'type_mock',
          link: new URL('https://link_mock'),
          checksum: 'checksum_mock',
          size: 0,
          extension: 'extension_mock',
          installer_link: installer ? new URL('https://installer_link_mock') : undefined,
          installer_checksum: installer ? 'installer_checksum_mock' : undefined,
          installer_size: installer ? 0 : undefined,
          installer_extension: installer ? 'installer_extension_mock' : undefined,
        },
      ],
    },
  },
})

export const createRandomContributorViewData = (): Contributor => ({
  avatarUri: 'avatarUri_mock',
  profileUri: 'profileUri_mock',
  login: 'login_mock',
  contributionsCount: 0,
  commitsListUri: 'commitsListUri_mock',
  repo: 'reponame_mock',
});

export const createRandomLatestForOSData = (): Binary => ({
  release_name: 'release_name_mock',
  link: new URL('https://link_mock'),
});

export const createRandomNewsAndEventsData = (): News => ({
  news: {
    news: [
      {
        id: 'id_mock',
        title: 'title_mock',
        body: 'body_mock',
        date: new Date(Date.UTC(2020, 0, 1)),
        link: new URL('https://link_mock'),
      },
    ],
    pagination: {
      page: 1,
      pagesize: 5,
      result_start: 1,
      result_end: 5,
      result_size: 5,
      total_result_size: 16,
    }
  },
  events: [
    {
      id: 'id_mock',
      title: 'title_mock',
      infoLink: new URL('https://infoLink_mock'),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ]
});

export const createRandomContributorViewData1Contribution = (): Contributor => ({
  avatarUri: 'avatarUri_mock',
  profileUri: 'profileUri_mock',
  login: 'login_mock',
  contributionsCount: 1,
  commitsListUri: 'commitsListUri_mock',
  repo: 'reponame_mock',
});