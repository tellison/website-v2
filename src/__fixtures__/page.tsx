import {
    AsciidocPage,
    File,
    MDXPage,
    SingleMDXPage,
    SiteMetaData
} from '../types';

export const createAsciidocData = (): {
    asciidoc: AsciidocPage;
    file: File;
  } => ({
    asciidoc: {
        id: 'asciidoc-1',
        html: `
            <p>Hello World, this is a <pre><code>test</code></pre></p>
            <table><thead><td>Header 1</td><td>Header 2</td></thead><tbody><tr><td>Cell 1</td><td class="icon">Cell 2</td></tr></tbody></table>
            <a href="https://www.eclipse.org">Eclipse</a>
            <span class="icon">[docker]</span>
            <span class="icon">[check]</span>
        `,
        document: {
            title: 'Asciidoc Page title',
            main: 'Asciidoc Page content',
        },
        pageAttributes: {
            authors: 'author1, author2',
        },
        fields: {
            slug: '/asciidoc/asciidoc-page-title',
        }
    },
    file: {
        relativePath: 'test.adoc',
    }
});

export const createMDXData = (): {
    allMdx: MDXPage;
} => ({
    allMdx: {
        edges: [
            {
                node: {
                    frontmatter: {
                        title: 'MDX Page title',
                        author: 'pmc',
                        date: '2021-01-01',
                        describtion: 'MDX Page describtion',
                    },
                    fields: {
                        slug: '/mdx/mdx-page-title',
                        postPath: '/mdx/mdx-page-title',
                    },
                    excerpt: 'MDX Page excerpt',
                }
            }
        ]
    }
});

export const createSingleMDXData = (): {
    mdx: SingleMDXPage;
    site: SiteMetaData;
} => ({
    mdx: {
        id: 'mdx-1',
        excerpt: 'MDX Page excerpt',
        frontmatter: {
            title: 'MDX Page title',
            author: 'pmc',
            date: '2021-01-01',
            description: 'MDX Page describtion',
            tags: ['tag1', 'tag2'],
        }
    },
    site: {
        siteMetadata: {
            title: 'Site title',
            siteUrl: 'https://adoptium.net',
            social: {
                twitter: 'twitter',
            }
        }
    }
});