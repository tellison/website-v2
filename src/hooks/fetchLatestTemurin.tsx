import { useEffect, useState } from 'react';
import { VersionMetaData } from ".";

const baseUrl = 'https://api.adoptium.net/v3';

export function fetchLatestForOS(
    isVisible: boolean,
    version: number,
    os: string,
    arch: string
): Binary | null {
    if (!os) {
        return null
    }
    const [binary, setBinary] = useState<Binary | null>(null);
    useEffect(() => {
        if (isVisible) {
        (async () => {
            setBinary(await fetchLatestForOSRequest(version, os, arch));
        })();
        }
    }, [isVisible]);

    return binary;
}

async function fetchLatestForOSRequest(version: number, os: string, arch: string) {
    const url = `${baseUrl}/assets/feature_releases/${version}/ga?os=${os}&architecture=${arch}&image_type=jdk&jvm_impl=hotspot&page_size=1&vendor=eclipse`;
    const response = await fetch(url);
    const json: LatestTemurin = (await response.json())[0];
    let binary_link = json.binaries[0].package.link
    if (json.binaries[0].installer) {
        binary_link = json.binaries[0].installer.link
    }
    return {
        release_name: json.release_name,
        link: binary_link
    };
}

export interface Binary {
  release_name: string;
  link: URL;
}

export interface LatestTemurin {
    download_count: number;
    id: string;
    release_link: URL;
    release_name: string;
    release_type: string;
    source?: {
        link: URL;
        name: string;
        size: number;
    }
    timestamp: Date;
    updated_at: Date;
    vendor: string;
    version_data: VersionMetaData
    binaries: [
        {
            architecture: string;
            download_count: number;
            heap_size: string;
            image_type: string;
            jvm_impl: string;
            os: string;
            package: {
                checksum: string;
                checksum_link: URL;
                download_count: number;
                link: URL;
                metadata_link: URL;
                name: string;
                signature_link: URL;
                size: number;
            }
            installer?: {
                checksum: string;
                checksum_link: URL;
                download_count: number;
                link: URL;
                metadata_link: URL;
                name: string;
                signature_link: URL;
                size: number;
            }
        }
    ]
}