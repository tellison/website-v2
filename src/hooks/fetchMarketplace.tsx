const baseUrl = 'https://marketplace-api.adoptium.net';
import { VersionMetaData } from ".";
    
export async function getAllPkgsForVersion(
    version: number,
    os: string,
    architecture: string,
    package_type: string,
    checkboxRef: any,
): Promise<MarketplaceRelease[] | null> {
    let microsoftSelected = checkboxRef.current.vendorMicrosoft.checked;
    let temurinSelected = checkboxRef.current.vendorAdoptium.checked;
    let redhatSelected = checkboxRef.current.vendorRedHat.checked;
    let huaweiSelected = checkboxRef.current.vendorHuawei.checked;
    let zuluSelected = checkboxRef.current.vendorAzul.checked;
    let ibmSelected = checkboxRef.current.vendorIBM.checked;
    let alibabaSelected = checkboxRef.current.vendorAlibaba.checked;

    let params = '?'
    params += 'feature_version=' + version;

    if (os !== 'any' ) {
        if (os === 'alpine-linux') {
            params += '&os=alpine_linux'
        } else {
            params += ('&os=' + os)
        }
    }

    if (architecture !== 'any' ) {
        params += ('&architecture=' + architecture)
    }

    if (package_type === 'any') {
        params += '&image_type=jdk&image_type=jre'
    } else {
        params += ('&image_type=' + package_type)
    }

    if (temurinSelected) {
        params += '&vendor=adoptium'
    }

    if (redhatSelected) {
        params += '&vendor=redhat'
    }

    if (huaweiSelected) {
        params += '&vendor=huawei'
    }

    if (microsoftSelected) {
        params += ('&vendor=microsoft')
    }

    if (zuluSelected) {
        params += ('&vendor=azul')
    }

    if (ibmSelected) {
        params += ('&vendor=ibm')
    }

    if (alibabaSelected) {
        params += ('&vendor=alibaba')
    }

    const url = new URL(baseUrl + '/v1/assets/latestForVendors' + params);
    const data = await getPkgs(url);
    return data
}

async function getPkgs(url: URL) {
    const response = await fetch(url)
    return response.json();
}

export function getImageForDistribution(distribution: string) {
    switch(distribution) {
        case 'temurin': return '/images/adoptium-logo.png';
        case 'redhat': return '/images/redhat.svg';
        case 'bisheng': return '/images/huawei.svg';
        case 'zulu': return '/images/azul-logo.png';
        case 'semeru': return '/images/ibm-logo.png';
        default: return `/images/${distribution}-logo.png`;
    }
}

export interface MarketplaceRelease {
    release_name: string;
    vendor: string;
    binary: {
        os: string;
        architecture: string;
        image_type: string;
        jvm_impl: string;
        package: {
            name: string;
            link: URL;
            sha265sum: string;
            sha256sum_link: URL;
            signature_link: URL;
        }
        installer?: [{
            name: string;
            link: URL;
            sha265sum: string;
            sha256sum_link: URL;
            signature_link: URL;
        }]
        timestamp: Date;
        scm_ref: string;
        openjdk_scm_ref: string;
        distribution: string;
        aqavit_results_link: URL;
        tck_affidavit_link: URL;
    }
    version: VersionMetaData;
}
