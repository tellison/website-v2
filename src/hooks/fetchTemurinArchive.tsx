import moment from 'moment';
import { VersionMetaData } from '.';
import { fetchExtension } from '../util/fetchExtension';

const baseUrl = 'https://api.adoptium.net/v3';

let releases: TemurinReleases[] = []

export async function getAssetsForVersion(
  version: number,
  releaseType: any,
  numBuilds: number,
  buildDate: Date,
  page: number
): Promise<ReturnedReleases | null> {
  let url = new URL(`${baseUrl}/assets/feature_releases/${version}/${releaseType}?vendor=eclipse`);
  url.searchParams.append('page', page.toString());
  if (numBuilds) {
    url.searchParams.append('page_size', numBuilds.toString());
  }
  if (buildDate) {
    url.searchParams.append('before', moment(buildDate).format('Y-MM-DD'));
  }
  // Expose total page count in header for pagination
  if (releaseType == 'ga') {
    url.searchParams.append('show_page_count', 'true');
  }
  releases = []
  const response = await getPkgs(url)
  const packages = await response.json()
  const pagecount = Number(response.headers.get('pagecount'))
  let pkgsFound: TemurinReleases[] = []
  for (let pkg of packages) {
      pkgsFound.push(pkg);
  }
  return renderReleases(pkgsFound, pagecount);
}

async function getPkgs(url: URL) {
  let response = await fetch(url)
  return response;
}

function renderReleases(pkgs, pagecount) {
  pkgs.forEach((aRelease) => {
    const release: TemurinReleases = {
        release_name: aRelease.release_name,
        release_link: aRelease.release_link,
        timestamp: aRelease.timestamp,
        platforms: {},
    };

    aRelease.binaries.forEach((
      aReleaseAsset: APIResponse) => {
        const platform = `${aReleaseAsset.os}-${aReleaseAsset.architecture}`
  
        // Skip this asset if it's not a binary type we're interested in displaying
        const binary_type = aReleaseAsset.image_type.toUpperCase();
        if (binary_type == 'SOURCES') {
          release.source_url = new URL(aReleaseAsset.package.link);
        }
        if (!['INSTALLER', 'JDK', 'JRE'].includes(binary_type)) {
          return;
        }
  
        if (!release.platforms[platform]) {
          release.platforms[platform] = {
            assets: [],
          };
        }
  
        let binary_constructor: ReleaseAsset = {
          os: aReleaseAsset.os,
          architecture: aReleaseAsset.architecture,
          type: binary_type,
          link: aReleaseAsset.package.link,
          checksum: aReleaseAsset.package.checksum,
          size: Math.floor(aReleaseAsset.package.size / 1000 / 1000),
          extension: fetchExtension(aReleaseAsset.package.name)
        };
  
        if (aReleaseAsset.installer) {
          binary_constructor.installer_link = aReleaseAsset.installer.link;
          binary_constructor.installer_checksum = aReleaseAsset.installer.checksum;
          binary_constructor.installer_size =  Math.floor(aReleaseAsset.installer.size / 1000 / 1000);
          binary_constructor.installer_extension = fetchExtension(aReleaseAsset.installer.name)
        }
  
        // Add the new binary to the release asset
        release.platforms[platform].assets.push(binary_constructor);
    });
    releases.push(release);
  })
  return {releases, pagecount}
}

export interface ReturnedReleases {
  releases: TemurinReleases[];
  pagecount: number;
}

export interface TemurinReleases {
  release_name: string;
  release_link: URL;
  source_url?: URL;
  timestamp: Date;
  platforms: {
    [key: string]: {
      assets: ReleaseAsset[];
    };
  };
}

interface ReleaseAsset {
  os: string;
  architecture: string;
  type: string;
  link: URL;
  checksum: string;
  size: number;
  extension: string;
  installer_link?: URL;
  installer_checksum?: string;
  installer_size?: number;
  installer_extension?: string;
}

interface APIResponse {
  os: string;
  architecture: string;
  image_type: string;
  package: {
    name: string;
    link: URL;
    checksum: string;
    size: number;
  };
  installer?: {
    link: URL;
    name: string;
    checksum: string;
    size: number;
  };
}

export interface MockTemurinFeatureReleaseAPI {
  id: string;
  download_count: number;
  release_name: string;
  release_type: string;
  release_link: URL;
  source?: {
    link: URL;
    name: string;
    size: number;
  };
  binaries: APIResponse[];
  timestamp: Date;
  updated_at: Date;
  vendor: string;
  version_data: VersionMetaData;
}