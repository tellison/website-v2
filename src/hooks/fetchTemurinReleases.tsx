import { VersionMetaData } from '.';
import { fetchExtension } from '../util/fetchExtension';

const baseUrl = 'https://api.adoptium.net/v3';

export async function loadLatestAssets(
    version: number,
    os: string,
    architecture: string,
    packageType: string
): Promise<ReleaseAsset[]> {
    let url = new URL(`${baseUrl}/assets/latest/${version}/hotspot?`);
    if (os !== 'any') {
        url.searchParams.append('os', os);
    }
    if (architecture !== 'any') {
        url.searchParams.append('architecture', architecture);
    }
    if (packageType !== 'any') {
        url.searchParams.append('image_type', packageType);
    }
    let data = await getPkgs(url);
    let pkgsFound: TemurinRelease[] = []
    for (let pkg of data) {
        pkgsFound.push(pkg);
    }
    return renderReleases(pkgsFound);
}

async function getPkgs(url: URL) {
    let response = await fetch(url)
    return response.json();
}

function renderReleases(pkgs: Array<TemurinRelease>): ReleaseAsset[] {
    let releases: ReleaseAsset[] = []
    pkgs.forEach((releaseAsset: TemurinRelease) => {
        const platform = `${releaseAsset.binary.os}-${releaseAsset.binary.architecture}`

        // Skip this asset if it's not a binary type we're interested in displaying
        const binary_type = releaseAsset.binary.image_type.toUpperCase();
        if (binary_type == 'SOURCES') {
            releases['source'] = releaseAsset;
        }
        if (!['INSTALLER', 'JDK', 'JRE'].includes(binary_type)) {
            return;
        }
        // Get the existing release asset (passed to the template) or define a new one
        let release: ReleaseAsset | undefined = releases.find((release: ReleaseAsset) => release.platform_name === platform);
        if (!release) {
            release = {
                platform_name: `${releaseAsset.binary.os}-${releaseAsset.binary.architecture}`,
                os: releaseAsset.binary.os,
                architecture: releaseAsset.binary.architecture,
                release_name: releaseAsset.release_name,
                release_link: new URL(releaseAsset.release_link),
                release_date: new Date(releaseAsset.binary.updated_at),
                binaries: []
            };
        }

        let binary_constructor: Binary = {
            type: binary_type,
            link: releaseAsset.binary.package.link,
            checksum: releaseAsset.binary.package.checksum,
            size: Math.floor(releaseAsset.binary.package.size / 1000 / 1000),
            extension: fetchExtension(releaseAsset.binary.package.name)
        };

        if (releaseAsset.binary.installer) {
            binary_constructor.installer_link = releaseAsset.binary.installer.link;
            binary_constructor.installer_checksum = releaseAsset.binary.installer.checksum;
            binary_constructor.installer_size =  Math.floor(releaseAsset.binary.installer.size / 1000 / 1000);
            binary_constructor.installer_extension = fetchExtension(releaseAsset.binary.installer.name);
        }

        // Add the new binary to the release asset
        release.binaries.push(binary_constructor);

        // We have the first binary, so add the release asset.
        if (release.binaries.length === 1) {
            releases.push(release);
        }

        releases = orderPlatforms(releases, 'platform_ordinal');
        releases.forEach((release) => {
            release.binaries.sort((binaryA, binaryB) => binaryA.type > binaryB.type ? 1 : binaryA.type < binaryB.type ? -1 : 0);
        });
    })
    return releases
}

function orderPlatforms (input, attr = 'thisPlatformOrder') {
  return sortByProperty(input, attr, false);
};

function sortByProperty (input, property, descending) {
  const invert = descending ? -1 : 1;
  const sorter = (a, b) => {
    return invert * (a[property] > b[property] ? 1 : a[property] < b[property] ? -1 : 0);
  };

  if (Array.isArray(input)) {
    return input.sort(sorter);
  } else {
    // Preserve the source object key as '_key'
    return Object.keys(input)
      .map(_key => Object.assign(input[_key], {_key}))
      .sort(sorter);
  }
};

export interface ReleaseAsset {
    platform_name: string;
    os: string;
    architecture: string;
    release_name: string;
    release_link: URL;
    release_date: Date;
    binaries: Array<Binary>;
}

interface TemurinRelease {
    release_link: URL;
    os: string;
    architecture: string;
    platform_name: string;
    release_name: string;
    release_date: Date;
    binary: {
        updated_at: Date;
        os: string;
        architecture: string;
        image_type: string;
        jvm_impl: string;
        package: {
            name: string;
            link: URL;
            checksum: string;
            checksum_link: URL;
            signature_link: URL;
            metadata_link: URL;
            size: number;
        }
        installer?: {
            name: string;
            link: URL;
            checksum: string;
            checksum_link: URL;
            signature_link: URL;
            metadata_link: URL;
            size: number;
        }
    }
}

interface Binary {
    type: string;
    link: URL;
    checksum: string;
    size: number;
    extension: string;
    installer_link?: URL;
    installer_checksum?: string;
    installer_size?: number;
    installer_extension?: string;
}

export interface MockTemurinReleaseAPI {
    release_link: URL;
    release_name: string;
    vendor: string;
    version: VersionMetaData
    binary: {
        project: string;
        scm_ref: string;
        updated_at: Date;
        os: string;
        architecture: string;
        download_count: number;
        heap_size: string;
        image_type: string;
        jvm_impl: string;
        package: {
            download_count: number;
            name: string;
            link: URL;
            checksum: string;
            checksum_link: URL;
            signature_link: URL;
            metadata_link: URL;
            size: number;
        }
        installer?: {
            download_count: number;
            name: string;
            link: URL;
            checksum: string;
            checksum_link: URL;
            signature_link: URL;
            metadata_link: URL;
            size: number;
        }
    }
}
