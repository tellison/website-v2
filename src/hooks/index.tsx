export * from './useOnScreen';
export * from './useAdoptiumContributorsApi';
export * from './fetchMarketplace';
export * from './fetchTemurinReleases';
export * from './fetchTemurinArchive';
export * from './fetchLatestTemurin';
export * from './fetchNews';

export interface VersionMetaData {
    major: number;
    minor: number;
    security: number;
    pre?: string;
    patch: number;
    build: number;
    optional?: string;
    openjdk_version: string;
}