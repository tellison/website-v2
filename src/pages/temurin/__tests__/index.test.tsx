import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest'
import { useOnScreen } from '../../../hooks/useOnScreen';
import { fetchLatestForOS } from '../../../hooks/fetchLatestTemurin';
import { createRandomLatestForOSData } from '../../../__fixtures__/hooks';
import { axe } from 'vitest-axe';
import Index, { Head } from '../index';

vi.mock('../../../hooks/useOnScreen');
vi.mock('../../../hooks/fetchLatestTemurin');

describe('Temurin Index page', () => {
  it('renders correctly', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    useOnScreen.mockReturnValue(true);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fetchLatestForOS.mockReturnValue(createRandomLatestForOSData());
    const { container } = render(<Index />);
    // eslint-disable-next-line
    const pageContent = container.querySelector('main');

    expect(pageContent).toMatchSnapshot();
  });

  it('head renders correctly', () => {
    const { container } = render(<Head />);
    // eslint-disable-next-line
    const title = container.querySelector('title');

    waitFor(() => {
      expect(title).toHaveTextContent('Temurin | Adoptium');
    });
  });

  it('has no accessibility violations', async () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    useOnScreen.mockReturnValue(true);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fetchLatestForOS.mockReturnValue(createRandomLatestForOSData());
    const { container } = render(<Index />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
