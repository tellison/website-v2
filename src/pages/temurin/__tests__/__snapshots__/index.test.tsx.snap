// Vitest Snapshot v1

exports[`Temurin Index page > renders correctly 1`] = `
<main>
  <section
    class="py-5 text-center container"
  >
    <div
      class="row py-lg-5"
    >
      <div
        class="col-lg-10 col-md-8 mx-auto"
      >
        <h1
          class="fw-light"
        >
          Eclipse Temurinâ„¢
        </h1>
        <div
          class="row align-items-center pt-3"
        >
          <div
            class="col-6 col-md-4"
          >
            <img
              alt="Temurin logo"
              class="img-fluid"
              src="/images/temurin-light.png"
              width="150"
            />
          </div>
          <div
            class="col-12 col-sm-6 col-md-8"
          >
            <p
              class="text-start"
            >
              The Eclipse Temurin project provides code and processes that support the building of runtime binaries and associated technologies that are high performance, enterprise-caliber, cross-platform, open-source licensed, and Java SE TCK-tested for general use across the Java ecosystem.
            </p>
          </div>
          <div
            class="container"
          >
            <h2
              class="fw-light mt-3 "
            >
              Text
            </h2>
            <div
              class="btn-group-vertical mx-auto col-6"
            >
              <a
                class="btn btn-lg btn-primary mt-3 py-3 text-white"
                href="/download"
                state="[object Object]"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
                  />
                </svg>
                 
                Text
                <br />
                <span
                  class="font-weight-light"
                  style="font-size: .6em;"
                >
                  release_name_mock
                </span>
              </a>
              <a
                class="btn btn-outline-dark mt-3"
                href="/temurin/releases"
              >
                Text
                 
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
                  />
                </svg>
              </a>
              <a
                class="btn btn-outline-dark mt-3"
                href="/temurin/archive"
              >
                Text
                 
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M32 448c0 17.7 14.3 32 32 32h384c17.7 0 32-14.3 32-32V160H32v288zm160-212c0-6.6 5.4-12 12-12h104c6.6 0 12 5.4 12 12v8c0 6.6-5.4 12-12 12H204c-6.6 0-12-5.4-12-12v-8zM480 32H32C14.3 32 0 46.3 0 64v48c0 8.8 7.2 16 16 16h480c8.8 0 16-7.2 16-16V64c0-17.7-14.3-32-32-32z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</main>
`;
